### PR TITLE
Make `deref_into_dyn_supertrait` lint the impl and not the usage

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Proposed by @compiler-errors in https://github.com/rust-lang/rust/issues/89460#issuecomment-1320806785
r? @crlf0710

### CODE REVIEW

File: compiler/rustc_lint/src/deref_into_dyn_supertrait.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            && let Some(target) = cx.get_associated_type(t, did, "Target")
+            && let ty::Dynamic(data, _, ty::Dyn) = target.kind()
+            && let Some(target_principal) = data.principal()
+            // `target_principal` is a supertrait of `t_principal`
+            && supertraits(cx.tcx, t_principal.with_self_ty(cx.tcx, cx.tcx.types.trait_object_dummy_self))
+                .any(|sup| sup.map_bound(|x| ty::ExistentialTraitRef::erase_self_ty(cx.tcx, x)) == target_principal)
+        {
+            cx.struct_span_lint(
+                DEREF_INTO_DYN_SUPERTRAIT,
+                item.span,
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/deref_into_dyn_supertrait.rs:75]: Can you use the `def_span` of either the `impl` or the associated item? It's a bit cleaner that way.

(or better yet, underline the impl and then put a label on the associated item)

**Comment by crlf0710** [comment]
I'm not in the code review good , so i need to reassign.

r? @compiler-errors

File: compiler/rustc_lint/src/deref_into_dyn_supertrait.rs (hunk 1 above)
Comment (WaffleLapkin) [reply compiler/rustc_lint/src/deref_into_dyn_supertrait.rs]: Like so?

Comment (compiler-errors) [reply compiler/rustc_lint/src/deref_into_dyn_supertrait.rs]: yes 😃

Comment (WaffleLapkin) [reply compiler/rustc_lint/src/deref_into_dyn_supertrait.rs]: well, I did :)

**Pull request merged**